lib/apollo-federation/schema.rb

```diff
@@ -12,49 +12,74 @@ def self.included(klass)
-            type.metadata[:federation_directives]&.any? { |directive| directive[:name] == 'key' }
-        end
+      def query(new_query_object = nil)
+        if new_query_object
+          @orig_query_object = new_query_object
+        else
+          if !@federation_query_object
+            # Build the new query object with the '_service' field
+            @federation_query_object = federation_query(@orig_query_object)
 
-        @query_object = federation_query
+            # If there are any "entities", define the Entity union and and the Query._entities field
+            possible_entities = schema_entities
+            if !possible_entities.empty?
+              entity_type = Class.new(Entity) do
+                possible_types(*possible_entities)
+              end
+              # TODO: Should/can we encapsulate all of this inside the module? What's the best/most
+              # Ruby way to split this out?
+              @federation_query_object.define_entities_field(entity_type)
+            end
 
-        if !possible_entities.empty?
-          entity_type = Class.new(Entity) do
-            possible_types(*possible_entities)
+            super(@federation_query_object)
           end
-          # TODO: Should/can we encapsulate all of this inside the module? What's the best/most Ruby
-          # way to split this out?
-          @query_object.define_entities_field(entity_type)
+
+          super
         end
+      end
+
+      def federation_sdl(context: nil)
+        document_from_schema = FederatedDocumentFromSchemaDefinition.new(self, context: context)
+        GraphQL::Language::Printer.new.print(document_from_schema.document)
+      end
+
+      private
+
+      def schema_entities
+        # Create a temporary schema that inherits from this one to extract the types
+        types_schema = Class.new(self)
+        # Add the original query objects to the types. We have to use orphan_types here to avoid
+        # infinite recursion
+        types_schema.orphan_types(@orig_query_object)
 
-        super
+        # Walk through all of the types and determine which ones are entities (any type with a
+        # "key" directive)
+        types_schema.types.values.select do |type|
+          # TODO: Interfaces can have a key...
+          !type.introspection? && type.include?(ApolloFederation::Object) &&
+            type.federation_directives&.any? { |directive| directive[:name] == 'key' }
+        end
       end
 
-      def federation_query
-        if query.nil?
+      def federation_query(query_obj)
+        if query_obj.nil?
           base = GraphQL::Schema::Object
         elsif Gem::Version.new(GraphQL::VERSION) >= Gem::Version.new('1.10.0')
-          base = query
+          base = query_obj
         else
-          base = query.metadata[:type_class]
+          # TODO: Will this ever happen?
+          base = query_obj.metadata[:type_class]
         end
 
         Class.new(base) do
```

Comment:
Wrapping the query type in another class can cause issues with other gems like this one https://github.com/exAspArk/graphql-guard#policy-object

What if we do it with class_eval ? My test suite broke because of this change.

```
base.class_eval do
  include EntitiesField
  include ServiceField
```

Reply:
Interesting. Is this something that's currently broken in v1.0.4? If so, I'd probably rather fix it in another PR/issue.

Reply:
Yes. Let's do it in a different PR!